aws/ci_build.rake

```diff
@@ -203,16 +197,18 @@ def upgrade_frontend(name, host)
     'git pull --ff-only',
     'sudo bundle install',
     'rake build',
+    # Use the AWS-provided script to cleanly re-register a frontend instance with its load balancer(s).
+    'bin/deploy_frontend/register_with_elb.sh'
   ]
   command = commands.join(' && ')
 
   HipChat.log "Upgrading <b>#{name}</b> (#{host})..."
 
   log_path = aws_dir "deploy-#{name}.log"
 
-  # Stop the frontend before running the commands so that the git pull doesn't modify files
-  # out from under a running instance. The rake build command will restart the instance.
-  stop_frontend name, host, log_path
+  # Remove the frontend from load balancer rotation before running the commands,
+  # so that the git pull doesn't modify files out from under a running instance.
+  deregister_frontend name, host, log_path
```

Comment:
Updated the code so that the initial call to `deregister_frontend` from `update_frontend` falls within the `begin` block, so that an error from this initial deregister command causes the `failed to upgrade` message to be reported to chat and the upgrade on that frontend to abort.

When the script call via `ssh` returns a non-zero exit code, which it returns when the deregistration fails for whatever reason, it will raise an error.

Existing code causes the whole deploy process to fail immediately (and noisily in chat) when the total number of frontend-deploy failures breaches the failure threshold (currently set to 5 instances). I agree that leaving our frontend fleet in an indeterminate state in the event of an upgrade failure is not ideal- until we move to immutable updates in the future, there is no way to fully prevent this possibility. (Implementing a rollback process on failure would be a slight improvement, but doesn't prevent the possibility that the rollback process itself will also fail).